Bug: no-restricted-imports shows an error only on last element when using an array of different importNames and same name

### Environment

Environment Info:

Node version: v15.10.0
npm version: v7.5.3
Local ESLint version: v7.27.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 20.6.0



### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
'use strict'

module.exports = {
  root: true,
  plugins: ['@typescript-eslint', 'react', 'jest', 'testing-library', 'prettier'],
  settings: {
    react: {
      pragma: 'React',
      version: '16.3'
    }
  },
  extends: [
    '@react-native-community',
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:jest/recommended',
    'plugin:testing-library/react',
    'prettier'
  ],
  env: {
    browser: true,
    commonjs: true,
    es6: true,
    node: true,
    jest: true
  },
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 7,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true,
      generators: true,
      impliedStrict: true,
      experimentalObjectRestSpread: true
    }
  },
  rules: {
    'react/prop-types': [0],
    'prettier/prettier': 'error',
    '@typescript-eslint/explicit-member-accessibility': 'off',
    '@typescript-eslint/no-floating-promises': 'off',
    'jest/no-try-expect': 'off',
    'jest/no-conditional-expect': 'off',
    'react-hooks/exhaustive-deps': 'warn',
    'no-redeclare': 'warn',
    'testing-library/prefer-screen-queries': 'off', //https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/prefer-screen-queries.md
    'testing-library/await-async-query': 'warn',
    /* Temporally inactive rules to be changed ultimately little by little */
    'jest/no-mocks-import': 'warn',
    'react/display-name': 'warn',
    'react/no-did-update-set-state': 'warn',
    'no-console': 'error',
    'no-restricted-imports': [
      'error',
      {
        paths: [
          { name: 'react-native', importNames: ['Text'], message: 'import Text from ui/_components instead ' },
          {
            name: 'react-native',
            importNames: ['TextInput'],
            message: 'import TextInput from ui/_components instead '
          },
          { name: 'react-native', importNames: ['View'], message: 'import View from ui/_components instead ' },
          {
            name: 'react-native',
            importNames: ['ScrollView'],
            message: 'import ScrollView from ui/_components instead '
          },
          {
            name: 'react-native',
            importNames: ['KeyboardAvoidingView'],
            message: 'import KeyboardAvoidingView from ui/_components instead '
          },
          {
            name: 'react-native',
            importNames: ['ImageBackground'],
            message: 'import ImageBackground from ui/_components instead '
          },
          {
            name: 'react-native',
            importNames: ['Image'],
            message: 'import Image from ui/_components instead '
          }
        ]
      }
    ]
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      rules: {
        '@typescript-eslint/explicit-member-accessibility': ['error', { accessibility: 'no-public' }],
        'no-unused-vars': 'off' //Causes duplicates with typescript-eslint/no-unused-vars
      }
    }
  ],
  globals: {
    process: true,
    fetchMock: true
  }
}

```
</details>

```js
import { Image , Text, ScrollView } from 'react-native'
```

Eslint only shows an error on Image as it is de last element of the array. If moving Text for example to the end of the array then it will show an error only on Text not on Image or ScrollView


### What did you expect to happen?

$ eslint --quiet --ext=js,jsx,ts,tsx ./src

/Users/user1/workspace/MyAPP/src/ui/_components/atoms/Icon/Icon.tsx
  2:10  error  'Image' import from 'react-native' is restricted. import Image from ui/_components instead        no-restricted-imports
  2:17  error  'ScrollView' import from 'react-native' is restricted. import ScrollView from ui/_components instead   no-restricted-imports
  2:29  error  'Text' import from 'react-native' is restricted. import Text from ui/_components instead         no-restricted-imports

✖ 3 problems (3 errors, 0 warnings)

error Command failed with exit code 1.


### What actually happened?

$ eslint --quiet --ext=js,jsx,ts,tsx ./src

/Users/user1/workspace/MyAPP/src/ui/_components/atoms/Icon/Icon.tsx
  2:10  error  'Image' import from 'react-native' is restricted. import Image from ui/_components instead   no-restricted-imports

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.


### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

--- a/docs/src/use/migrate-to-9.0.0.md
+++ b/docs/src/use/migrate-to-9.0.0.md
@@ -30,6 +30,7 @@ The lists below are ordered roughly by the number of users each change is expect
 * [New checks in `no-implicit-coercion` by default](#no-implicit-coercion)
 * [Case-sensitive flags in `no-invalid-regexp`](#no-invalid-regexp)
 * [`varsIgnorePattern` option of `no-unused-vars` no longer applies to catch arguments](#vars-ignore-pattern)
+* [`no-restricted-imports` now accepts multiple config entries with the same `name`](#no-restricted-imports)
 * [`"eslint:recommended"` and `"eslint:all"` strings no longer accepted in flat config](#string-config)
 * [`no-inner-declarations` has a new default behavior with a new option](#no-inner-declarations)
 
@@ -280,6 +281,41 @@ try {
 **To address:** If you want to specify ignore patterns for `catch` clause variable names, use the `caughtErrorsIgnorePattern` option in addition to `varsIgnorePattern`.
 
 **Related issue(s):** [#17540](https://github.com/eslint/eslint/issues/17540)
+
+## <a name="no-restricted-imports"></a> `no-restricted-imports` now accepts multiple config entries with the same `name`
+
+In previous versions of ESLint, if multiple entries in the `paths` array of your configuration for the `no-restricted-imports` rule had the same `name` property, only the last one would apply, while the previous ones would be ignored.
+
+As of ESLint v9.0.0, all entries apply, allowing for specifying different messages for different imported names. For example, you can now configure the rule like this:
+
+```js
+{
+    rules: {
+        "no-restricted-imports": ["error", {
+            paths: [
+                {
+                    name: "react-native",
+                    importNames: ["Text"],
+                    message: "import 'Text' from 'ui/_components' instead"
+                },
+                {
+                    name: "react-native",
+                    importNames: ["View"],
+                    message: "import 'View' from 'ui/_components' instead"
+                }
+            ]
+        }]
+    }
+}
+```
+
+and both `import { Text } from "react-native"` and `import { View } from "react-native"` will be reported, with different messages.
+
+In previous versions of ESLint, with this configuration only `import { View } from "react-native"` would be reported.
+
+**To address:** If your configuration for this rule has multiple entries with the same `name`, you may need to remove unintentional ones.
+
+**Related issue(s):** [#15261](https://github.com/eslint/eslint/issues/15261)
 
 ## <a name="string-config"></a> `"eslint:recommended"` and `"eslint:all"` no longer accepted in flat config
 

--- a/lib/rules/no-restricted-imports.js
+++ b/lib/rules/no-restricted-imports.js
@@ -161,17 +161,25 @@ module.exports = {
             (Object.hasOwn(options[0], "paths") || Object.hasOwn(options[0], "patterns"));
 
         const restrictedPaths = (isPathAndPatternsObject ? options[0].paths : context.options) || [];
-        const restrictedPathMessages = restrictedPaths.reduce((memo, importSource) => {
+        const groupedRestrictedPaths = restrictedPaths.reduce((memo, importSource) => {
+            const path = typeof importSource === "string"
+                ? importSource
+                : importSource.name;
+
+            if (!memo[path]) {
+                memo[path] = [];
+            }
+
             if (typeof importSource === "string") {
-                memo[importSource] = { message: null };
+                memo[path].push({});
             } else {
-                memo[importSource.name] = {
+                memo[path].push({
                     message: importSource.message,
                     importNames: importSource.importNames
-                };
+                });
             }
             return memo;
-        }, {});
+        }, Object.create(null));
 
         // Handle patterns too, either as strings or groups
         let restrictedPatterns = (isPathAndPatternsObject ? options[0].patterns : []) || [];
@@ -203,57 +211,59 @@ module.exports = {
          * @private
          */
         function checkRestrictedPathAndReport(importSource, importNames, node) {
-            if (!Object.hasOwn(restrictedPathMessages, importSource)) {
+            if (!Object.hasOwn(groupedRestrictedPaths, importSource)) {
                 return;
             }
 
-            const customMessage = restrictedPathMessages[importSource].message;
-            const restrictedImportNames = restrictedPathMessages[importSource].importNames;
+            groupedRestrictedPaths[importSource].forEach(restrictedPathEntry => {
+                const customMessage = restrictedPathEntry.message;
+                const restrictedImportNames = restrictedPathEntry.importNames;
 
-            if (restrictedImportNames) {
-                if (importNames.has("*")) {
-                    const specifierData = importNames.get("*")[0];
+                if (restrictedImportNames) {
+                    if (importNames.has("*")) {
+                        const specifierData = importNames.get("*")[0];
 
+                        context.report({
+                            node,
+                            messageId: customMessage ? "everythingWithCustomMessage" : "everything",
+                            loc: specifierData.loc,
+                            data: {
+                                importSource,
+                                importNames: restrictedImportNames,
+                                customMessage
+                            }
+                        });
+                    }
+
+                    restrictedImportNames.forEach(importName => {
+                        if (importNames.has(importName)) {
+                            const specifiers = importNames.get(importName);
+
+                            specifiers.forEach(specifier => {
+                                context.report({
+                                    node,
+                                    messageId: customMessage ? "importNameWithCustomMessage" : "importName",
+                                    loc: specifier.loc,
+                                    data: {
+                                        importSource,
+                                        customMessage,
+                                        importName
+                                    }
+                                });
+                            });
+                        }
+                    });
+                } else {
                     context.report({
                         node,
-                        messageId: customMessage ? "everythingWithCustomMessage" : "everything",
-                        loc: specifierData.loc,
+                        messageId: customMessage ? "pathWithCustomMessage" : "path",
                         data: {
                             importSource,
-                            importNames: restrictedImportNames,
                             customMessage
                         }
                     });
                 }
-
-                restrictedImportNames.forEach(importName => {
-                    if (importNames.has(importName)) {
-                        const specifiers = importNames.get(importName);
-
-                        specifiers.forEach(specifier => {
-                            context.report({
-                                node,
-                                messageId: customMessage ? "importNameWithCustomMessage" : "importName",
-                                loc: specifier.loc,
-                                data: {
-                                    importSource,
-                                    customMessage,
-                                    importName
-                                }
-                            });
-                        });
-                    }
-                });
-            } else {
-                context.report({
-                    node,
-                    messageId: customMessage ? "pathWithCustomMessage" : "path",
-                    data: {
-                        importSource,
-                        customMessage
-                    }
-                });
-            }
+            });
         }
 
         /**

--- a/tests/lib/rules/no-restricted-imports.js
+++ b/tests/lib/rules/no-restricted-imports.js
@@ -1020,6 +1020,199 @@ ruleTester.run("no-restricted-imports", rule, {
             endColumn: 18
         }]
     },
+
+    // https://github.com/eslint/eslint/issues/15261
+    {
+        code: "import { Image, Text, ScrollView } from 'react-native'",
+        options: [{
+            paths: [
+                {
+                    name: "react-native",
+                    importNames: ["Text"],
+                    message: "import Text from ui/_components instead"
+                },
+                {
+                    name: "react-native",
+                    importNames: ["TextInput"],
+                    message: "import TextInput from ui/_components instead"
+                },
+                { name: "react-native", importNames: ["View"], message: "import View from ui/_components instead " },
+                {
+                    name: "react-native",
+                    importNames: ["ScrollView"],
+                    message: "import ScrollView from ui/_components instead"
+                },
+                {
+                    name: "react-native",
+                    importNames: ["KeyboardAvoidingView"],
+                    message: "import KeyboardAvoidingView from ui/_components instead"
+                },
+                {
+                    name: "react-native",
+                    importNames: ["ImageBackground"],
+                    message: "import ImageBackground from ui/_components instead"
+                },
+                {
+                    name: "react-native",
+                    importNames: ["Image"],
+                    message: "import Image from ui/_components instead"
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "'Image' import from 'react-native' is restricted. import Image from ui/_components instead",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 10,
+                endColumn: 15
+            },
+            {
+                message: "'Text' import from 'react-native' is restricted. import Text from ui/_components instead",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 17,
+                endColumn: 21
+            },
+            {
+                message: "'ScrollView' import from 'react-native' is restricted. import ScrollView from ui/_components instead",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 23,
+                endColumn: 33
+            }
+        ]
+    },
+    {
+        code: "import { foo, bar, baz } from 'mod'",
+        options: [{
+            paths: [
+                {
+                    name: "mod",
+                    importNames: ["foo"],
+                    message: "Import foo from qux instead."
+                },
+                {
+                    name: "mod",
+                    importNames: ["baz"],
+                    message: "Import baz from qux instead."
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "'foo' import from 'mod' is restricted. Import foo from qux instead.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 10,
+                endColumn: 13
+            },
+            {
+                message: "'baz' import from 'mod' is restricted. Import baz from qux instead.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 20,
+                endColumn: 23
+            }
+        ]
+    },
+    {
+        code: "import * as mod from 'mod'",
+        options: [{
+            paths: [
+                {
+                    name: "mod",
+                    importNames: ["foo"],
+                    message: "Import foo from qux instead."
+                },
+                {
+                    name: "mod",
+                    importNames: ["bar"],
+                    message: "Import bar from qux instead."
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "* import is invalid because 'foo' from 'mod' is restricted. Import foo from qux instead.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 8,
+                endColumn: 16
+            },
+            {
+                message: "* import is invalid because 'bar' from 'mod' is restricted. Import bar from qux instead.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 8,
+                endColumn: 16
+            }
+        ]
+    },
+    {
+        code: "import { foo } from 'mod'",
+        options: [{
+            paths: [
+
+                // restricts importing anything from the module
+                {
+                    name: "mod"
+                },
+
+                // message for a specific import name
+                {
+                    name: "mod",
+                    importNames: ["bar"],
+                    message: "Import bar from qux instead."
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "'mod' import is restricted from being used.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 1,
+                endColumn: 26
+            }
+        ]
+    },
+    {
+        code: "import { bar } from 'mod'",
+        options: [{
+            paths: [
+
+                // restricts importing anything from the module
+                {
+                    name: "mod"
+                },
+
+                // message for a specific import name
+                {
+                    name: "mod",
+                    importNames: ["bar"],
+                    message: "Import bar from qux instead."
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "'mod' import is restricted from being used.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 1,
+                endColumn: 26
+            },
+            {
+                message: "'bar' import from 'mod' is restricted. Import bar from qux instead.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 10,
+                endColumn: 13
+            }
+        ]
+    },
+
     {
         code: "import foo, { bar } from 'mod';",
         options: [{

--- a/tests/lib/rules/no-restricted-imports.js
+++ b/tests/lib/rules/no-restricted-imports.js
@@ -1117,6 +1117,91 @@ ruleTester.run("no-restricted-imports", rule, {
         ]
     },
     {
+        code: "import { foo, bar, baz, qux } from 'mod'",
+        options: [{
+            paths: [
+                {
+                    name: "mod",
+                    importNames: ["bar"],
+                    message: "Use `barbaz` instead of `bar`."
+                },
+                {
+                    name: "mod",
+                    importNames: ["foo", "qux"],
+                    message: "Don't use 'foo' and `qux` from 'mod'."
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "'foo' import from 'mod' is restricted. Don't use 'foo' and `qux` from 'mod'.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 10,
+                endColumn: 13
+            },
+            {
+                message: "'bar' import from 'mod' is restricted. Use `barbaz` instead of `bar`.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 15,
+                endColumn: 18
+            },
+            {
+                message: "'qux' import from 'mod' is restricted. Don't use 'foo' and `qux` from 'mod'.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 25,
+                endColumn: 28
+            }
+        ]
+    },
+    {
+        code: "import { foo, bar, baz, qux } from 'mod'",
+        options: [{
+            paths: [
+                {
+                    name: "mod",
+                    importNames: ["foo", "baz"],
+                    message: "Don't use 'foo' or 'baz' from 'mod'."
+                },
+                {
+                    name: "mod",
+                    importNames: ["a", "c"],
+                    message: "Don't use 'a' or 'c' from 'mod'."
+                },
+                {
+                    name: "mod",
+                    importNames: ["b", "bar"],
+                    message: "Use 'b' or `bar` from 'quux/mod' instead."
+                }
+            ]
+        }],
+        errors: [
+            {
+                message: "'foo' import from 'mod' is restricted. Don't use 'foo' or 'baz' from 'mod'.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 10,
+                endColumn: 13
+            },
+            {
+                message: "'bar' import from 'mod' is restricted. Use 'b' or `bar` from 'quux/mod' instead.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 15,
+                endColumn: 18
+            },
+            {
+                message: "'baz' import from 'mod' is restricted. Don't use 'foo' or 'baz' from 'mod'.",
+                type: "ImportDeclaration",
+                line: 1,
+                column: 20,
+                endColumn: 23
+            }
+        ]
+    },
+    {
         code: "import * as mod from 'mod'",
         options: [{
             paths: [

